subprojects/common/http-client/src/main/kotlin/com/avito/http/internal/StatsHttpEventListener.kt

```diff
@@ -47,30 +48,36 @@ internal class StatsHttpEventListener(
     }
 
     private fun serviceMetric(request: Request): SeriesName {
-        val metadata = request.tag(RequestMetadata::class.java)
+        return requestMetadataProvider.provide(request).fold(
+            onSuccess = { metadata ->
+                SeriesName.create(
+                    "service",
+                    metadata.serviceName,
+                    metadata.methodName
+                )
+            },
+            onFailure = { throwable ->
+                val urlWithoutParams = request.url.newBuilder()
+                    .query(null)
+                    .build()
 
-        return if (metadata == null) {
-            val urlWithoutParams = request.url.newBuilder()
-                .query(null)
-                .build()
+                val seriesName = SeriesName.create(
+                    "service",
+                    "unknown-service",
+                    "unknown-method"
+                )
 
-            logger.warn(
-                msg = "RequestMetadata not available for: $urlWithoutParams\n" +
-                    "You should add okhttp.Request.tag() with additional request information for metrics\n" +
-                    "or use Retrofit @Tag annotation",
-                error = null
-            )
-            SeriesName.create(
-                "service",
-                "unknown-service",
-                "unknown-method"
-            )
-        } else {
-            SeriesName.create(
-                "service",
-                metadata.serviceName,
-                metadata.methodName
-            )
-        }
+                logger.warn(
```

Comment:
How about fail-fast? Is it possible?

Reply:
What will happen if we throw an exception? Will it be swallowed or something else?

Reply:
i don't know
at least it will break contract of upstream functions